### PR TITLE
CDDL:babbage: Switch to using Huddle/Cuddle

### LINKED ISSUE #4781: Babbage: Generate CDDL using Huddle/Cuddle
Just like `Shelley`, `Allegra`, `Mary`, `Alonzo`, and `Conway` are already doing it.

### PR DESCRIPTION
**Pull request by aniketd**
# Description

- Remove
  - crypto.cddl
  - extras.cddl
- Add
  - Test.Cardano.Ledger.Babbage.CDDL
  - cardano-ledger-babbage:exe:huddle-cddle
  - huddleRoundTrip(Ann)CborSpec for all definitions already being tested
- Autogenerate babbage.cddl using Huddle/Cuddle
- Update
  - hie.yaml with babbage paths module and huddle-cddl
  - gen-cddl.sh to include babbage

Resolves #4781 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs
Comment (teodanciu) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs]: This was the case for alonzo, but it's not for babbage, no? 
```suggestion

```

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs
Comment (teodanciu) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs]: Not new in babbage
```suggestion
```

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs
Comment (teodanciu) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs]: These are not new for babbage 
```suggestion
        |16: min pool cost
        |17: ada per utxo byte
        |18: cost models for script languages
        |19: execution costs
        |20: max tx ex units
        |21: max block ex units
        |22: max value size
        |23: collateral percentage
        |24: max collateral inputs
```

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs
Comment (teodanciu) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs:111]: There should be a tuple here inside the array, is this what your FIXME comment above refers to? 
I reckon we can't merge this before it's fixed

Is it difficult?  Isn't it just a matter of having the array over the tuple that was defined in alonzo? `grp [a major_protocol_version, a VUInt]`   Or is there more to the problem?

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs
Comment (teodanciu) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs:95]: Is this why in the generated cddl, `operational_cert` is defined as  `($kes_vkey, uint, uint, $signature)` instead of how it was before, 
```
  [ hot_vkey        : $kes_vkey
  , sequence_number : uint
  , kes_period      : uint
  , sigma           : $signature
  ]
```

because it is defined in a way that enables reusing it like this, but which atm means that the name of the keys are lost ? Is this right?  Sorry I know we talked about this on the previous PR but I didn't really understand why that was happening, so wondering if my conclusion is correct

**Review by teodanciu — COMMENTED** [review]
Looks good to me, went through the cddl differences and found a few fields which were incorrectly indicated as NEW in the comment, and a difference which is more significant in the way the protocol_version is defined, and the latter I believe we need to fix before merging. 
Nice work!

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs
Comment (aniketd) [reply eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs]:
```suggestion
        |NEW:
        |  3: [* plutus_v2_script]
```

`plutus_v2_script` is new at index 3. Thanks for the catch. I have added it :raised_hands:

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs
Comment (aniketd) [reply eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs:111]: It is a stale comment because I just repeated what is done in `conway`. From what I read, this is because CBOR doesn't have a direct way to represent a "group" as defined by CDDL parentheses. The parentheses in CDDL are a syntactic grouping mechanism rather than a data structure specification. It is very confusing actually, to use a group as the element of a 1-element list. The tests pass because the two definitions are equivalent for _our_ purposes, though they are actually different in theory.

Specifically, groups delimited by parentheses are also just encoded as arrays in CBOR. So `[(a,b)]` will be encoded as `array of an array containing a and b`, whereas, `[a,b]` or `(a, b)` both will be encoded as `array containing a and b`.

Please correct me if I'm wrong @lehins.

Comment (teodanciu) [reply eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs:111]: I see. Interesting, thanks!

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs
Comment (aniketd) [reply eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs:95]: Yes that is correct. Using an `array []` vs using a `group ()`, we loose keys in the latter. Thanks for bringing this up! This just made me think, we can safely replace `groups` with `arrays` and have no problems with the encoding. I have updated the definition and all the `.cddl` files. Please let me know what you think.

Comment (aniketd) [reply eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/CDDL.hs:95]: It appears that in the case of the `OCert` we can't do this because of the way its CBOR instance is defined. I'm reverting back to the old way, and yes, we will loose the keys for documentation, but I will add them as a comment. :+1:

**Review by teodanciu — APPROVED** [review]
Looks good to me

**Pull request merged**
